File: src/bfs/bus-routes.py
from collections import defaultdict

class Solution:
    def num_buses_to_destination(self, routes, s, t):
        to_routes = defaultdict(set)
        for i, route in enumerate(routes):
            for j in route:
                to_routes[j].add(i)
        
        bfs = [(s, 0)]
        visited = set([s])

        for stop, bus in bfs:
            if stop == t:
                return bus
            for i in to_routes[stop]:
                for j in routes[i]:
                    if j not in visited:
                        bfs.append([j, bus + 1])
                        visited.add(j)
                routes[i] = []
        return -1
        

File: src/bfs/bus_routes.rs
use crate::Solution;
use std::collections::{HashMap, HashSet};

impl Solution {
    pub fn num_buses_to_destination(routes: Vec<Vec<i32>>, source: i32, target: i32) -> i32 {
        let mut map = HashMap::<i32, HashSet<i32>>::new();
        for (i, route) in routes.iter().enumerate() {
            let ii32 = i as i32;
            for j in route {
                map.entry(*j).or_insert(HashSet::new()).insert(ii32);
            }
        }
        let mut bfs = vec![[source, 0]];
        let mut seen = HashSet::new();
        seen.insert(source);
        while !bfs.is_empty() {
            let [cur, step] = bfs.remove(0);
            if cur == target {
                return step;
            }
            if let Some(routes) = map.get(&cur) {
                for route in routes {
                    if seen.contains(route) {
                        continue;
                    }
                    seen.insert(*route);
                    for stop in routes {
                        if !seen.contains(stop) {
                            bfs.push([*stop, step + 1]);
                        }
                    }
                }
            }
        }
        -1
    }
}
